components/blitz/src/omero/gateway/facility/DataManagerFacility.java

```diff
@@ -497,4 +572,242 @@ public void addImagesToDataset(SecurityContext ctx,
+                try {
+                    OriginalFile originalFile = new OriginalFileI();
+                    originalFile.setName(omero.rtypes.rstring(name));
+                    originalFile.setPath(omero.rtypes.rstring(path));
+                    originalFile.setSize(omero.rtypes.rlong(file.length()));
+                    final ChecksumAlgorithm checksumAlgorithm = new ChecksumAlgorithmI();
+                    checksumAlgorithm.setValue(omero.rtypes
+                            .rstring(ChecksumAlgorithmSHA1160.value));
+                    originalFile.setHasher(checksumAlgorithm);
+                    originalFile.setMimetype(omero.rtypes.rstring(mime));
+                    originalFile = (OriginalFile) saveAndReturnObject(ctx,
+                            originalFile);
+
+                    rawFileStore = gateway.getRawFileService(ctx);
+                    rawFileStore.setFileId(originalFile.getId().getValue());
+                    FileInputStream stream = new FileInputStream(file);
+                    long pos = 0;
+                    int rlen;
+                    byte[] buf = new byte[INC];
+                    ByteBuffer bbuf;
+                    while (((rlen = stream.read(buf)) > 0) && !cb.isCancelled()) {
+                        rawFileStore.write(buf, pos, rlen);
+                        pos += rlen;
+                        bbuf = ByteBuffer.wrap(buf);
+                        bbuf.limit(rlen);
+                    }
+                    stream.close();
+
+                    if (cb.isCancelled()) {
+                        try {
+                            rawFileStore.close();
+                        } catch (ServerError e) {
+                        }
+                        cb.setResult(null);
+                        return;
+                    }
+
+                    originalFile = rawFileStore.save();
+
+                    FileAnnotation fa = new FileAnnotationI();
+                    fa.setFile(originalFile);
+                    if (description != null)
+                        fa.setDescription(omero.rtypes.rstring(description));
+                    fa.setNs(omero.rtypes.rstring(namespace));
+                    fa = (FileAnnotation) saveAndReturnObject(ctx, fa);
+
+                    if (target != null)
+                        cb.setResult(attachAnnotation(ctx,
+                                new FileAnnotationData(fa), target));
+                    else
+                        cb.setResult(new FileAnnotationData(fa));
+                } catch (Throwable t) {
+                    cb.setException(t);
+                } finally {
+                    if (rawFileStore != null) {
+                        try {
+                            rawFileStore.close();
+                        } catch (ServerError e) {
+                        }
+                    }
+                }
+            }
+        };
+
+        if (callback != null) {
+            (new Thread(r)).start();
```

Comment:
I'd urge against using `new Thread` anywhere that it's not absolutely necessary. If that means we need to look into a `ThreadingFacility` or a `BackgroundService` etc. :+1: